R/ChromBackend.R

```diff
@@ -796,5 +828,87 @@ setMethod("split", "ChromBackend", function(x, f, drop = FALSE, ...) {
 ################################################################################
 ## Filter functions
 
-## TODO:
-## - generic filterChromData and filterPeaksData
+#' @exportMethod filterChromData
+#'
+#' @note
+#' Filter chromData removes chromatograms that do not match the filter criteria.
+#' FilterPeaks Data does not, only some data points are removed.
+#'
+#' can i make it possible that it can filter both character and numeric columns?
+#' mm idk for now.. i feel like for character values just using base R function makes sense,
+#' can i make it so that it keeps what i want or remove?
+#' @rdname ChromBackend
+#'
+#' @importFrom MsCoreUtils between
+#'
+setMethod("filterChromData", "ChromBackend",
+          function(object, variables = character(),
+                   ranges = numeric(), match = c("any", "all"),
+                   keep = TRUE) {
+              if (!length(chromVariables) || !length(ranges))
+                  return(object)
+              if (!is.numeric(ranges))
+                  stop("filterChromData only support filtering for numerical ",
+                       "'variables'")
+              match <- match.arg(match)
+              if (is.character(variables)){
+                  if(!all(variables %in% chromVariables(object)))
+                      stop("One or more values passed with parameter ",
+                           "'variables' are not available as chromatogram ",
+                           "variables in object. Use the 'chromVariables()' ",
+                           "function to list possible values.")
+              } else
+                  stop("The 'variables' parameter needs to be of type ",
+                       "'character'.")
+              if (length(variables) != length(ranges) / 2)
+                  stop("Length of 'ranges' needs to be twice the length of ",
+                       "the parameter 'variables' and define the lower ",
+                       "and upper bound for values of each chromatogram variable ",
+                       "defined with parameter 'variables'.")
+              query <- chromData(object, columns = variables)
+              idx <- .filter_ranges(query, ranges, match)
+              if (keep) object <- object[idx]
+              else object <- object[-idx]
+              object
+          })
+
+#' @exportMethod filterPeaksData
+#'
+#' @rdname ChromBackend
+#'
+#' @export
+setMethod("filterPeaksData", "ChromBackend",
```

Comment:
the function is great. My only concern: do we want to apply this filter immediately or do we want to do it similar to `Spectra`, i.e. using a processing queue in the `Chromatograms` object to cache this operation?

The main reason we might want that is for *read only* backends, i.e. such as SQL database(s), backends reading on-the-fly from mzML or similar files.

Reply:
I think we could have this method only implemented for `Chromatograms()` actually, not for the backend. All the checking for input parameters could be done there and then we simply add the `.filter_ranges()` function to the processing queue. Or we could have a parameter `lazy = FALSE` (or maybe better `immediate = TRUE`?) that would bypass the processing queue and apply the filter immediately? Let's discuss in the next meeting.

Reply:
yesss I think that makes sense.. Some chromatograms object are going to be big. I think it's good to prepare for that already.